Feat: user preferences backend

joshmhanson opened the pull request:
### Description

This adds `POST /api/users` with optional queries `timeZone`, `questionTypes`, `questionDifficulty`, `codingLanguage`. It uses a new type UserPreferences which is equivalent to the User type minus id, email, name, and imageUrl. I'm not sure if there's a cleaner way to do this with Express and TypeScript.

### Testing

As of yet this is untested.

### Checklist

- [x] I provided a description of the contents of this pull request.
- [ ] I've linked this pull request to an issue.
- [x] I've assigned myself as the Assignee and labelled the PR correctly.

joshmhanson committed:
feat: User preferences type and update function

joshmhanson committed:
feat: users endpoint and controller

wilsonplau commented on controllers/userController.ts:
```diff
@@ -0,0 +1,36 @@
+
+        try {
+            if (!timeZone && !questionTypes && !questionDifficulty && !codingLanguage)
+                return res.status(400).send({ error: "Empty Request." });
```
I probably wouldn't respond with an error here - I just would make sure it doesn't update anything and send the User back.

wilsonplau commented on controllers/userController.ts:
```diff
@@ -0,0 +1,36 @@
+
+            if (!userId) return res.status(401).send({ error: "Not Authenticated." });
+            await userRepository.updateUserById(userId, prefs);
+            res.send(200).json(prefs);
```
I would probably send a fully updated user back to the client so it can replace the User they have.

wilsonplau commented on models/UserRepository.ts:
```diff
@@ -24,4 +26,8 @@ export default class UserRepository {
   }
+
+  async updateUserById(id: string, preferences: Partial<UserPreferences>): Promise<void> {
+    await this.prisma.user.update({ where: { id }, data: { ...preferences } });
```
[nit]: I don't think preferences actually needs to be spread here.

wilsonplau reviewed (commented):
Code is very clean and great. Just one comment on the response.

joshmhanson committed:
fix: Return updated user in userController

Switched to using POST body rather than URL queries

joshmhanson committed:
fix: User is possibly null in updateUserById

joshmhanson commented:
Made the following updates:
- I switched to using a POST body rather than URL queries, as the latter would get awkward when a lot is being updated or when the values are arrays. Also greatly simplified the code.
- I changed `data: { ...preferences }` to just `data: preferences`, hopefully this works. Didn't give me a type error.
- `updateUserById` now returns `Promise<User | null>` so that the user controller can receive the updated user back from the function

joshmhanson committed:
fix: Throw error in updateUserById when user not found

joshmhanson commented:
`updateUserById` now returns `Promise<User>` (no null) and throws an error when user is not found.

joshmhanson committed:
fix: Remove unused imports from UserRepository.ts

wilsonplau reviewed (approved):
LGTM.
